internal/app/jwt.go

```diff
@@ -2,223 +2,74 @@ package app
+func jwtEchoMiddleware(cfg *ServerConfig) echo.MiddlewareFunc {
 
+	issuerURL, err := url.Parse("https://" + cfg.Config.Auth0.Domain)
 	if err != nil {
-		return nil, err
+		log.Fatalf("failed to parse the issuer url: %v", err)
 	}
 
-	return jso.jwks, nil
-}
-
-func fetchJwks(publicKeyURL string) ([]JSONWebKeys, error) {
-	resp, err := http.Get(publicKeyURL)
-	var res struct {
-		Jwks []JSONWebKeys `json:"keys"`
-	}
+	provider := jwks.NewCachingProvider(issuerURL, 60*time.Minute)
 
+	// Set up the validator.
+	jwtValidator, err := validator.New(
+		provider.KeyFunc,
+		validator.RS256,
+		cfg.Config.AuthClient.ISS[0],
+		cfg.Config.AuthClient.AUD,
+	)
 	if err != nil {
-		return nil, err
+		log.Fatalf("failed to set up the validator: %v", err)
 	}
-	defer func() {
-		_ = resp.Body.Close()
-	}()
 
-	err = json.NewDecoder(resp.Body).Decode(&res)
+	middleware := jwtmiddleware.New(jwtValidator.ValidateToken)
 
-	if err != nil {
-		return nil, err
-	}
-
-	return res.Jwks, nil
+	return echo.WrapMiddleware(middleware.CheckJWT)
 }
 
-func getPemCert(token *jwt.Token, publicKeyURL string, jwks Jwks) (string, error) {
-	cert := ""
-	keys, err := jwks.GetJwks(publicKeyURL)
-
-	if err != nil {
-		return cert, err
-	}
-
-	for k := range keys {
-		if token.Header["kid"] == keys[k].Kid {
-			cert = "-----BEGIN CERTIFICATE-----\n" + keys[k].X5c[0] + "\n-----END CERTIFICATE-----"
-		}
-	}
-
-	if cert == "" {
-		err := errors.New("unable to find appropriate key")
-		return cert, err
-	}
-
-	return cert, nil
-}
-
-func parseJwtMiddleware(cfg *ServerConfig) echo.MiddlewareFunc {
-	iss := urlFromDomain(cfg.Config.Auth0.Domain)
-	aud := cfg.Config.Auth0.Audience
-
+// load claim from ctx and inject the user sub into ctx
+func parseJwtMiddleware() echo.MiddlewareFunc {
 	return func(next echo.HandlerFunc) echo.HandlerFunc {
 		return func(c echo.Context) error {
 			req := c.Request()
 			ctx := req.Context()
 
-			token := ctx.Value(userProfileKey)
-			if userProfile, ok := token.(*jwt.Token); ok {
-				claims := userProfile.Claims.(jwt.MapClaims)
-
-				// Verify 'iss' claim
-				checkIss := claims.VerifyIssuer(iss, false)
-				if !checkIss {
-					return errorResponse(c, "invalid issuer")
-				}
-
-				// Verify 'aud' claim
-				if !verifyAudience(claims, aud) {
-					return errorResponse(c, "invalid audience")
-				}
+			rawClaims := ctx.Value(jwtmiddleware.ContextKey{})
+			if claims, ok := rawClaims.(*validator.ValidatedClaims); ok {
 
 				// attach sub and access token to context
-				if sub, ok := claims["sub"].(string); ok {
-					ctx = context.WithValue(ctx, contextAuth0Sub, sub)
-				}
+				ctx = context.WithValue(ctx, contextAuth0Sub, claims.RegisteredClaims.Subject)
+
+				/* this claim is sent from auth0
 				if user, ok := claims["https://reearth.io/user_id"].(string); ok {
 					ctx = context.WithValue(ctx, contextUser, user)
-				}
-				ctx = context.WithValue(ctx, contextAuth0AccessToken, userProfile.Raw)
-			}
+				}*/
```

Comment:
Please delete comments that simply comment out the code, as they cause confusion. We use Git, so we can browse past code at any time.